src/app-layer-ftp.c

```diff
@@ -312,49 +421,32 @@ static int FTPGetLine(FtpState *state)
 
 /**
  * \brief This function is called to determine and set which command is being
- * transfered to the ftp server
- * \param ftp_state the ftp state structure for the parser
+ * transferred to the ftp server
  * \param input input line of the command
  * \param len of the command
+ * \param cmd_descriptor when the command has been parsed
  *
  * \retval 1 when the command is parsed, 0 otherwise
  */
-static int FTPParseRequestCommand(void *ftp_state, uint8_t *input,
-                                  uint32_t input_len)
+static int FTPParseRequestCommand(uint8_t *input, uint32_t input_len, const FtpCommand **cmd_descriptor)
 {
     SCEnter();
-    FtpState *fstate = (FtpState *)ftp_state;
-    fstate->command = FTP_COMMAND_UNKNOWN;
-
-    if (input_len >= 4 && SCMemcmpLowercase("port", input, 4) == 0) {
-        fstate->command = FTP_COMMAND_PORT;
-    }
-
-    if (input_len >= 4 && SCMemcmpLowercase("eprt", input, 4) == 0) {
-        fstate->command = FTP_COMMAND_EPRT;
-    }
 
-    if (input_len >= 8 && SCMemcmpLowercase("auth tls", input, 8) == 0) {
-        fstate->command = FTP_COMMAND_AUTH_TLS;
-    }
+    *cmd_descriptor = NULL;
 
-    if (input_len >= 4 && SCMemcmpLowercase("pasv", input, 4) == 0) {
-        fstate->command = FTP_COMMAND_PASV;
-    }
-
-    if (input_len > 5 && SCMemcmpLowercase("retr", input, 4) == 0) {
-        fstate->command = FTP_COMMAND_RETR;
-    }
-
-    if (input_len >= 4 && SCMemcmpLowercase("epsv", input, 4) == 0) {
-        fstate->command = FTP_COMMAND_EPSV;
-    }
+    for (int i = 0; i < FTP_COMMAND_MAX; i++) {
+        if (!FtpCommands[i].command_length) {
+            break;
+        }
+        if (input_len >= FtpCommands[i].command_length &&
```

Comment:
Is not there some multi pattern matching algorithm already implemented in Suricata ? (using a dictionary structure)
This comes at the cost of some precomputation and memory but it saves many `SCMemcmpLowercase` at runtime

Reply:
Good question -- @victorjulien is there?

Reply:
Yes, have a look at how the smtp parser takes the string command and uses mpm to get its id.

Reply:
Issue 3077 created for command parsing performance improvement.